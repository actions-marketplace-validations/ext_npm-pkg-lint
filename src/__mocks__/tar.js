const tar = jest.createMockFromModule("tar");
let mockFilelist;

function list(options) {
	for (const filename of mockFilelist) {
		options.onentry({
			path: filename,
		});
	}
	return Promise.resolve();
}

function __setMockFiles(filelist) {
	mockFilelist = filelist;
}

tar.__setMockFiles = __setMockFiles;
tar.list = list;

module.exports = tar;
