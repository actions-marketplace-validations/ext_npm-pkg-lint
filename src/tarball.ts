import tar, { FileStat } from "tar";
import PackageJson from "./types/package-json";
import { isBlacklisted } from "./blacklist";
import { Message } from "./message";
import { Result } from "./result";

interface RequiredFile {
	field: string;
	ruleId: string;
	filename: string;
}

export async function getFileList(filename: string): Promise<string[]> {
	const entries: FileStat[] = [];
	await tar.list({
		file: filename,
		strict: true,
		onentry: (entry: FileStat) => entries.push(entry),
	});
	return entries.map((entry) => {
		const filename = (entry.path as unknown) as string;
		return filename.replace(/^package\//, "");
	});
}

export async function blacklistedFiles(filelist: string[]): Promise<string[]> {
	return filelist.filter(isBlacklisted);
}

function normalizeRequiredFiles(src: string | string[] | Record<string, string>): string[] {
	if (typeof src === "string") {
		return [src];
	} else if (Array.isArray(src)) {
		return src;
	} else {
		return Object.values(src);
	}
}

function* yieldRequiredFiles(
	src: string | string[] | Record<string, string>,
	template: Pick<RequiredFile, "field" | "ruleId">
): Generator<RequiredFile> {
	const files = normalizeRequiredFiles(src);
	for (const filename of files) {
		yield { ...template, filename };
	}
}

function* requiredFiles(pkg: PackageJson): Generator<RequiredFile> {
	if (pkg.main) {
		yield* yieldRequiredFiles(pkg.main, { field: "main", ruleId: "no-missing-main" });
	}
	if (pkg.browser) {
		yield* yieldRequiredFiles(pkg.browser, { field: "browser", ruleId: "no-missing-main" });
	}
	if (pkg.module) {
		yield* yieldRequiredFiles(pkg.module, { field: "module", ruleId: "no-missing-main" });
	}
	if (pkg["jsnext:main"]) {
		yield* yieldRequiredFiles(pkg["jsnext:main"], {
			field: "jsnext:main",
			ruleId: "no-missing-main",
		});
	}
	if (pkg.typings) {
		yield* yieldRequiredFiles(pkg.typings, { field: "typings", ruleId: "no-missing-typings" });
	}
	if (pkg.bin) {
		yield* yieldRequiredFiles(pkg.bin, { field: "bin", ruleId: "no-missing-binary" });
	}
	if (pkg.man) {
		yield* yieldRequiredFiles(pkg.man, { field: "man", ruleId: "no-missing-man" });
	}
}

export async function verifyTarball(pkg: PackageJson, filePath: string): Promise<Result> {
	const messages: Message[] = [];
	const filelist = await getFileList(filePath);

	for (const filename of await blacklistedFiles(filelist)) {
		messages.push({
			ruleId: "no-disallowed-files",
			severity: 2,
			message: `${filename} is not allowed in tarball`,
			line: 1,
			column: 1,
		});
	}

	for (const requiredFile of requiredFiles(pkg)) {
		if (!filelist.includes(requiredFile.filename)) {
			messages.push({
				ruleId: requiredFile.ruleId,
				severity: 2,
				message: `${requiredFile.filename} (pkg.${requiredFile.field}) is not present in tarball`,
				line: 1,
				column: 1,
			});
		}
	}

	return {
		messages,
		filePath,
		errorCount: messages.length,
		warningCount: 0,
	};
}
