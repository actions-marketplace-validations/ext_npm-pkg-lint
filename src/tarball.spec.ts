/* eslint-disable @typescript-eslint/no-var-requires */

jest.mock("tar");

import { verifyTarball } from "./tarball";
import PackageJson from "./types/package-json";

beforeEach(() => {
	require("tar").__setMockFiles([]);
});

it("should return error if disallowed file is found", async () => {
	expect.assertions(1);
	require("tar").__setMockFiles(["foo.spec.js"]);
	const pkg: PackageJson = {
		name: "mock-pkg",
		version: "1.2.3",
	};
	const result = await verifyTarball(pkg, "mock-path");
	expect(result.messages).toMatchInlineSnapshot(`
		Array [
		  Object {
		    "column": 1,
		    "line": 1,
		    "message": "foo.spec.js is not allowed in tarball",
		    "ruleId": "no-disallowed-files",
		    "severity": 2,
		  },
		]
	`);
});

describe("should return error if package.json references missing file", () => {
	it.each`
		field               | template
		${"main"}           | ${{ main: "index.js" }}
		${"browser"}        | ${{ browser: "index.js" }}
		${"module"}         | ${{ module: "index.js" }}
		${"jsnext:main"}    | ${{ "jsnext:main": "index.js" }}
		${"typings"}        | ${{ typings: "index.d.ts" }}
		${"bin (single)"}   | ${{ bin: "index.js" }}
		${"bin (multiple)"} | ${{ bin: { foo: "dist/foo.js", bar: "dist/bar.js" } }}
		${"man (single)"}   | ${{ man: "man/foo.1" }}
		${"man (multiple)"} | ${{ man: ["man/foo.1", "man/bar.1"] }}
	`("$field", async ({ template }) => {
		expect.assertions(1);
		require("tar").__setMockFiles([]);
		const pkg: PackageJson = {
			name: "mock-pkg",
			version: "1.2.3",
			...template,
		};
		const result = await verifyTarball(pkg, "mock-path");
		expect(result.messages).toMatchSnapshot();
	});
});

describe("should not return error if package.json references existing file", () => {
	it.each`
		field               | template
		${"main"}           | ${{ main: "index.js" }}
		${"browser"}        | ${{ browser: "index.js" }}
		${"module"}         | ${{ module: "index.js" }}
		${"jsnext:main"}    | ${{ "jsnext:main": "index.js" }}
		${"typings"}        | ${{ typings: "index.d.ts" }}
		${"bin (single)"}   | ${{ bin: "index.js" }}
		${"bin (multiple)"} | ${{ bin: { foo: "dist/foo.js", bar: "dist/bar.js" } }}
		${"man (single)"}   | ${{ man: "man/foo.1" }}
		${"man (multiple)"} | ${{ man: ["man/foo.1", "man/bar.1"] }}
	`("$field", async ({ template }) => {
		expect.assertions(1);
		require("tar").__setMockFiles([
			"index.js",
			"index.d.ts",
			"dist/foo.js",
			"dist/bar.js",
			"man/foo.1",
			"man/bar.1",
		]);
		const pkg: PackageJson = {
			name: "mock-pkg",
			version: "1.2.3",
			...template,
		};
		const result = await verifyTarball(pkg, "mock-path");
		expect(result.messages).toEqual([]);
	});
});
