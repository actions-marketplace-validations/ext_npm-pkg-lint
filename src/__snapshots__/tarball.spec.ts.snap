// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return error if package.json references missing file bin (multiple) 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "dist/foo.js (pkg.bin) is not present in tarball",
    "ruleId": "no-missing-binary",
    "severity": 2,
  },
  Object {
    "column": 1,
    "line": 1,
    "message": "dist/bar.js (pkg.bin) is not present in tarball",
    "ruleId": "no-missing-binary",
    "severity": 2,
  },
]
`;

exports[`should return error if package.json references missing file bin (single) 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "index.js (pkg.bin) is not present in tarball",
    "ruleId": "no-missing-binary",
    "severity": 2,
  },
]
`;

exports[`should return error if package.json references missing file browser 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "index.js (pkg.browser) is not present in tarball",
    "ruleId": "no-missing-main",
    "severity": 2,
  },
]
`;

exports[`should return error if package.json references missing file jsnext:main 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "index.js (pkg.jsnext:main) is not present in tarball",
    "ruleId": "no-missing-main",
    "severity": 2,
  },
]
`;

exports[`should return error if package.json references missing file main 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "index.js (pkg.main) is not present in tarball",
    "ruleId": "no-missing-main",
    "severity": 2,
  },
]
`;

exports[`should return error if package.json references missing file man (multiple) 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "man/foo.1 (pkg.man) is not present in tarball",
    "ruleId": "no-missing-man",
    "severity": 2,
  },
  Object {
    "column": 1,
    "line": 1,
    "message": "man/bar.1 (pkg.man) is not present in tarball",
    "ruleId": "no-missing-man",
    "severity": 2,
  },
]
`;

exports[`should return error if package.json references missing file man (single) 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "man/foo.1 (pkg.man) is not present in tarball",
    "ruleId": "no-missing-man",
    "severity": 2,
  },
]
`;

exports[`should return error if package.json references missing file module 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "index.js (pkg.module) is not present in tarball",
    "ruleId": "no-missing-main",
    "severity": 2,
  },
]
`;

exports[`should return error if package.json references missing file typings 1`] = `
Array [
  Object {
    "column": 1,
    "line": 1,
    "message": "index.d.ts (pkg.typings) is not present in tarball",
    "ruleId": "no-missing-typings",
    "severity": 2,
  },
]
`;
